// nylas/lib/esm/resources/credentials.js
import { Resource } from './resource.js';
import { makePathParams } from '../utils.js';
export class Credentials extends Resource {
    /**
     * Return all credentials
     * @return A list of credentials
     */
    list({ provider, queryParams, overrides, }) {
        return super._list({
            queryParams,
            overrides,
            path: makePathParams('/v3/connectors/{provider}/creds', { provider }),
        });
    }
    /**
     * Return a credential
     * @return The credential
     */
    find({ provider, credentialsId, overrides, }) {
        return super._find({
            path: makePathParams('/v3/connectors/{provider}/creds/{credentialsId}', {
                provider,
                credentialsId,
            }),
            overrides,
        });
    }
    /**
     * Create a credential
     * @return The created credential
     */
    create({ provider, requestBody, overrides, }) {
        return super._create({
            path: makePathParams('/v3/connectors/{provider}/creds', { provider }),
            requestBody,
            overrides,
        });
    }
    /**
     * Update a credential
     * @return The updated credential
     */
    update({ provider, credentialsId, requestBody, overrides, }) {
        return super._update({
            path: makePathParams('/v3/connectors/{provider}/creds/{credentialsId}', {
                provider,
                credentialsId,
            }),
            requestBody,
            overrides,
        });
    }
    /**
     * Delete a credential
     * @return The deleted credential
     */
    destroy({ provider, credentialsId, overrides, }) {
        return super._destroy({
            path: makePathParams('/v3/connectors/{provider}/creds/{credentialsId}', {
                provider,
                credentialsId,
            }),
            overrides,
        });
    }
}
