// nylas/lib/esm/resources/grants.js
import { Resource } from './resource.js';
import { makePathParams } from '../utils.js';
/**
 * Nylas Grants API
 *
 * The Nylas Grants API allows for the management of grants.
 */
export class Grants extends Resource {
    /**
     * Return all Grants
     * @return The list of Grants
     */
    async list({ overrides, queryParams } = {}, 
    /**
     * @deprecated Use `queryParams` instead.
     */
    _queryParams) {
        return super._list({
            queryParams: queryParams ?? _queryParams ?? undefined,
            path: makePathParams('/v3/grants', {}),
            overrides: overrides ?? {},
        });
    }
    /**
     * Return a Grant
     * @return The Grant
     */
    find({ grantId, overrides, }) {
        return super._find({
            path: makePathParams('/v3/grants/{grantId}', { grantId }),
            overrides,
        });
    }
    /**
     * Update a Grant
     * @return The updated Grant
     */
    update({ grantId, requestBody, overrides, }) {
        return super._updatePatch({
            path: makePathParams('/v3/grants/{grantId}', { grantId }),
            requestBody,
            overrides,
        });
    }
    /**
     * Delete a Grant
     * @return The deletion response
     */
    destroy({ grantId, overrides, }) {
        return super._destroy({
            path: makePathParams('/v3/grants/{grantId}', { grantId }),
            overrides,
        });
    }
}
