// nylas/lib/cjs/resources/bookings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bookings = void 0;
const utils_js_1 = require("../utils.js");
const resource_js_1 = require("./resource.js");
class Bookings extends resource_js_1.Resource {
    /**
     * Return a Booking
     * @return The booking
     */
    find({ bookingId, queryParams, overrides, }) {
        return super._find({
            path: (0, utils_js_1.makePathParams)('/v3/scheduling/bookings/{bookingId}', {
                bookingId,
            }),
            queryParams,
            overrides,
        });
    }
    /**
     * Create a Booking
     * @return The created booking
     */
    create({ requestBody, queryParams, overrides, }) {
        return super._create({
            path: (0, utils_js_1.makePathParams)('/v3/scheduling/bookings', {}),
            requestBody,
            queryParams,
            overrides,
        });
    }
    /**
     * Confirm a Booking
     * @return The confirmed Booking
     */
    confirm({ bookingId, requestBody, queryParams, overrides, }) {
        return super._update({
            path: (0, utils_js_1.makePathParams)('/v3/scheduling/bookings/{bookingId}', {
                bookingId,
            }),
            requestBody,
            queryParams,
            overrides,
        });
    }
    /**
     * Reschedule a Booking
     * @return The rescheduled Booking
     */
    reschedule({ bookingId, requestBody, queryParams, overrides, }) {
        return super._updatePatch({
            path: (0, utils_js_1.makePathParams)('/v3/scheduling/bookings/{bookingId}', {
                bookingId,
            }),
            requestBody,
            queryParams,
            overrides,
        });
    }
    /**
     * Delete a Booking
     * @return The deleted Booking
     */
    destroy({ bookingId, requestBody, queryParams, overrides, }) {
        return super._destroy({
            path: (0, utils_js_1.makePathParams)('/v3/scheduling/bookings/{bookingId}', {
                bookingId,
            }),
            requestBody,
            queryParams,
            overrides,
        });
    }
}
exports.Bookings = Bookings;
