// nylas/lib/cjs/resources/connectors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Connectors = void 0;
const resource_js_1 = require("./resource.js");
const credentials_js_1 = require("./credentials.js");
const utils_js_1 = require("../utils.js");
class Connectors extends resource_js_1.Resource {
    /**
     * @param apiClient client The configured Nylas API client
     */
    constructor(apiClient) {
        super(apiClient);
        this.credentials = new credentials_js_1.Credentials(apiClient);
    }
    /**
     * Return all connectors
     * @return A list of connectors
     */
    list({ queryParams, overrides, }) {
        return super._list({
            queryParams,
            overrides,
            path: (0, utils_js_1.makePathParams)('/v3/connectors', {}),
        });
    }
    /**
     * Return a connector
     * @return The connector
     */
    find({ provider, overrides, }) {
        return super._find({
            path: (0, utils_js_1.makePathParams)('/v3/connectors/{provider}', { provider }),
            overrides,
        });
    }
    /**
     * Create a connector
     * @return The created connector
     */
    create({ requestBody, overrides, }) {
        return super._create({
            path: (0, utils_js_1.makePathParams)('/v3/connectors', {}),
            requestBody,
            overrides,
        });
    }
    /**
     * Update a connector
     * @return The updated connector
     */
    update({ provider, requestBody, overrides, }) {
        return super._update({
            path: (0, utils_js_1.makePathParams)('/v3/connectors/{provider}', { provider }),
            requestBody,
            overrides,
        });
    }
    /**
     * Delete a connector
     * @return The deleted connector
     */
    destroy({ provider, overrides, }) {
        return super._destroy({
            path: (0, utils_js_1.makePathParams)('/v3/connectors/{provider}', { provider }),
            overrides,
        });
    }
}
exports.Connectors = Connectors;
