// nylas/lib/cjs/resources/credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Credentials = void 0;
const resource_js_1 = require("./resource.js");
const utils_js_1 = require("../utils.js");
class Credentials extends resource_js_1.Resource {
    /**
     * Return all credentials
     * @return A list of credentials
     */
    list({ provider, queryParams, overrides, }) {
        return super._list({
            queryParams,
            overrides,
            path: (0, utils_js_1.makePathParams)('/v3/connectors/{provider}/creds', { provider }),
        });
    }
    /**
     * Return a credential
     * @return The credential
     */
    find({ provider, credentialsId, overrides, }) {
        return super._find({
            path: (0, utils_js_1.makePathParams)('/v3/connectors/{provider}/creds/{credentialsId}', {
                provider,
                credentialsId,
            }),
            overrides,
        });
    }
    /**
     * Create a credential
     * @return The created credential
     */
    create({ provider, requestBody, overrides, }) {
        return super._create({
            path: (0, utils_js_1.makePathParams)('/v3/connectors/{provider}/creds', { provider }),
            requestBody,
            overrides,
        });
    }
    /**
     * Update a credential
     * @return The updated credential
     */
    update({ provider, credentialsId, requestBody, overrides, }) {
        return super._update({
            path: (0, utils_js_1.makePathParams)('/v3/connectors/{provider}/creds/{credentialsId}', {
                provider,
                credentialsId,
            }),
            requestBody,
            overrides,
        });
    }
    /**
     * Delete a credential
     * @return The deleted credential
     */
    destroy({ provider, credentialsId, overrides, }) {
        return super._destroy({
            path: (0, utils_js_1.makePathParams)('/v3/connectors/{provider}/creds/{credentialsId}', {
                provider,
                credentialsId,
            }),
            overrides,
        });
    }
}
exports.Credentials = Credentials;
